File: go/bob/bob.go
package bob

import (
	"strings"
	"unicode"
)

const testVersion = 3

func Hey(sentence string) string {
	if containsOnlySpaces(sentence) {
		return "Fine. Be that way!"
	} else if containsLetters(sentence) && containsOnlyUpperLetters(sentence) {
		return "Whoa, chill out!"
	} else if strings.HasSuffix(strings.TrimSpace(sentence), "?") {
		return "Sure."
	}
	return "Whatever."
}

func containsOnlySpaces(sentence string) bool {
	for _, r := range sentence {
		if !unicode.IsSpace(r) {
			return false
		}
	}
	return true
}

func containsLetters(sentence string) bool {
	for _, r := range sentence {
		if unicode.IsLetter(r) {
			return true
		}
	}
	return false
}

func containsOnlyUpperLetters(sentence string) bool {
	for _, r := range sentence {
		if unicode.IsLetter(r) && unicode.IsLower(r) {
			return false
		}
	}
	return true
}

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(declare contains-letter? is-upper-cased? ends-with-qst? does)

(defn response-for [s]
  (let [sentence (str/trim s)]
    (cond
      (does sentence [contains-letter? is-upper-cased? ends-with-qst?])
      "Calm down, I know what I'm doing!"
      (does sentence [contains-letter? is-upper-cased?])
      "Whoa, chill out!"
      (ends-with-qst? sentence)
      "Sure."
      (str/blank? sentence)
      "Fine. Be that way!"
      :else
      "Whatever.")))

(defn contains-letter? [s]
  (some? (some #(Character/isLetter %) s)))

(defn is-upper-cased? [s]
  (= (str/upper-case s) s))

(defn ends-with-qst? [s]
  (= \? (last s)))

(defn does [s, requirements]
  (every? #(% s) requirements))
